### ISSUE TITLE
Twitter filter harvests stop collecting

### ISSUE THREAD

**Issue by justinlittman** [gwu-libraries/sfm-ui#420]
On at least 2 occasions, twitter filter streams running in production have stopped collecting.  They are still running, but spinning on errors (see below).

**Comment by justinlittman** [comment]
Here's the relevant log from a failure in production:

```
2016-09-04 01:00:43,107: __main__ --> Processed 20700 tweets
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/SocketServer.py", line 599, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/local/lib/python2.7/SocketServer.py", line 334, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/local/lib/python2.7/site-packages/warcprox/warcprox.py", line 158, in __init__
    warcprox.mitmproxy.MitmProxyHandler.__init__(self, request, client_address, server)
  File "/usr/local/lib/python2.7/site-packages/warcprox/mitmproxy.py", line 25, in __init__
    http_server.BaseHTTPRequestHandler.__init__(self, request, client_address, server)
  File "/usr/local/lib/python2.7/SocketServer.py", line 655, in __init__
    self.handle()
  File "/usr/local/lib/python2.7/BaseHTTPServer.py", line 340, in handle
    self.handle_one_request()
  File "/usr/local/lib/python2.7/BaseHTTPServer.py", line 328, in handle_one_request
    method()
  File "/usr/local/lib/python2.7/site-packages/warcprox/mitmproxy.py", line 96, in do_CONNECT
    self.handle_one_request()
  File "/usr/local/lib/python2.7/BaseHTTPServer.py", line 328, in handle_one_request
    method()
  File "/usr/local/lib/python2.7/site-packages/warcprox/mitmproxy.py", line 131, in do_COMMAND
    self._proxy_request()
  File "/usr/local/lib/python2.7/site-packages/warcprox/warcprox.py", line 210, in _proxy_request
    buf = h.read(8192)
  File "/usr/local/lib/python2.7/httplib.py", line 588, in read
    return self._read_chunked(amt)
  File "/usr/local/lib/python2.7/httplib.py", line 630, in _read_chunked
    line = self.fp.readline(_MAXLINE + 1)
  File "/usr/local/lib/python2.7/site-packages/warcprox/warcprox.py", line 125, in readline
    hunk = self.fp.readline(size)
  File "/usr/local/lib/python2.7/socket.py", line 480, in readline
    data = self._sock.recv(self._rbufsize)
  File "/usr/local/lib/python2.7/ssl.py", line 734, in recv
    return self.read(buflen)
  File "/usr/local/lib/python2.7/ssl.py", line 621, in read
    v = self._sslobj.read(len or 1024)
SSLError: ('The read operation timed out',)
2016-09-04 01:01:48,857: root --> ('Connection broken: IncompleteRead(0 bytes read)', IncompleteRead(0 bytes read))
2016-09-04 01:01:48,857: root --> sleeping 1
2016-09-04 01:01:49,858: root --> connecting to filter stream for {'track': u'election2016, election, clinton, kaine, trump, pence', 'follow': u'1339835893, 25073877, 22203756, 17
2858784', 'stall_warning': True}
2016-09-04 01:01:49,861: requests.packages.urllib3.connectionpool --> Resetting dropped connection: stream.twitter.com
2016-09-04 01:01:49,861: requests.packages.urllib3.connectionpool --> Starting new HTTPS connection (2): stream.twitter.com
2016-09-04 01:01:59,886: root --> caught connection error HTTPSConnectionPool(host='stream.twitter.com', port=443): Max retries exceeded with url: /1.1/statuses/filter.json (Caused by ProxyError('Cannot connect to proxy.', error('Tunnel connection failed: 500 [Errno -2] Name or service not known',)))
2016-09-04 01:01:59,901: root --> closing existing http session
2016-09-04 01:01:59,901: root --> closing last response
2016-09-04 01:01:59,902: root --> creating http session
2016-09-04 01:01:59,904: requests.packages.urllib3.connectionpool --> Starting new HTTPS connection (1): stream.twitter.com
2016-09-04 01:02:09,917: root --> caught connection error HTTPSConnectionPool(host='stream.twitter.com', port=443): Max retries exceeded with url: /1.1/statuses/filter.json (Cause
d by ProxyError('Cannot connect to proxy.', error('Tunnel connection failed: 500 [Errno -2] Name or service not known',)))
2016-09-04 01:02:09,917: root --> closing existing http session
2016-09-04 01:02:09,917: root --> closing last response
2016-09-04 01:02:09,918: root --> creating http session
2016-09-04 01:02:09,919: requests.packages.urllib3.connectionpool --> Starting new HTTPS connection (1): stream.twitter.com
2016-09-04 01:02:19,935: root --> caught connection error HTTPSConnectionPool(host='stream.twitter.com', port=443): Max retries exceeded with url: /1.1/statuses/filter.json (Cause
d by ProxyError('Cannot connect to proxy.', error('Tunnel connection failed: 500 [Errno -2] Name or service not known',)))
2016-09-04 01:02:19,936: root --> closing existing http session
2016-09-04 01:02:19,936: root --> closing last response
2016-09-04 01:02:19,936: root --> creating http session
2016-09-04 01:02:19,938: requests.packages.urllib3.connectionpool --> Starting new HTTPS connection (1): stream.twitter.com
2016-09-04 01:02:29,951: root --> caught connection error HTTPSConnectionPool(host='stream.twitter.com', port=443): Max retries exceeded with url: /1.1/statuses/filter.json (Cause
d by ProxyError('Cannot connect to proxy.', error('Tunnel connection failed: 500 [Errno -2] Name or service not known',)))
2016-09-04 01:02:29,952: root --> closing existing http session
2016-09-04 01:02:29,952: root --> closing last response
2016-09-04 01:02:29,952: root --> creating http session
```

**Comment by justinlittman** [comment]
Also:

```
2016-09-06 07:41:53,914 864 ERROR Thread-9584 warcprox.warcprox.WarcProxyHandler.log_error(mitmproxy.py:143) localhost - - [06/Sep/2016 07:41:53] code 500, message [Errno -3] Temporary failure in name resolution
2016-09-06 07:41:53,915 864 INFO Thread-9584 warcprox.warcprox.WarcProxyHandler.log_message(mitmproxy.py:147) WarcProxyHandler localhost - - [06/Sep/2016 07:41:53] "CONNECT stream.twitter.com:443 HTTP/1.0" 500 -
2016-09-06 07:41:53,917 864 WARNING Thread-9584 warcprox.warcprox.WarcProxyHandler.do_CONNECT(mitmproxy.py:91) failed to send error response ([Errno -3] Temporary failure in name resolution) to proxy client: [Errno 32] Broken pipe
```

**Comment by justinlittman** [comment]
Interesting.  This happens at 1am on a Sunday night.  I wonder if there is something that happens at that particular time in WRLC infrastructure.

**Comment by justinlittman** [comment]
Speculation is that this is either something in WRLC infrastructure or a problem with Twitter blocking the streams.

Regardless, twarc's retry strategy is not successfully re-establishing a connection.  One possible solution is a fail-slow strategy. This means that the error will be rethrown/not caught if it has occurred a certain number of times (within Twarc).  This number will be configurable.

**Comment by justinlittman** [comment]
To test thesis that something in WRLC infrastructure, running election filter stream in production (WRLC) and test (AWS).  Now lets see if they both stop and/or stop at the same time.

**Comment by justinlittman** [comment]
In sfm-test, seeing this:
```
2016-09-11 08:43:49,448: requests.packages.urllib3.connectionpool --> Starting new HTTPS connection (1): stream.twitter.com
2016-09-11 08:43:49,494: root --> caught connection error ("bad handshake: Error([('rsa routines', 'RSA_padding_check_PKCS1_type_1', 'block type is not 01'), ('rsa routines', 'RSA_EAY_PUBLIC_DECRYPT', 'padding check failed'), ('asn1 encoding routines', 'ASN1_item_verify', 'EVP lib'), ('SSL routines', 'SSL3_GET_SERVER_CERTIFICATE', 'certificate verify failed')],)",)
2016-09-11 08:43:49,494: root --> closing existing http session
2016-09-11 08:43:49,494: root --> creating http session
2016-09-11 08:43:49,497: requests.packages.urllib3.connectionpool --> Starting new HTTPS connection (1): stream.twitter.com
2016-09-11 08:43:49,543: root --> caught connection error ("bad handshake: Error([('rsa routines', 'RSA_padding_check_PKCS1_type_1', 'block type is not 01'), ('rsa routines', 'RSA_EAY_PUBLIC_DECRYPT', 'padding check failed'), ('asn1 encoding routines', 'ASN1_item_verify', 'EVP lib'), ('SSL routines', 'SSL3_GET_SERVER_CERTIFICATE', 'certificate verify failed')],)",)
2016-09-11 08:43:49,543: root --> closing existing http session
2016-09-11 08:43:49,544: root --> creating http session
```

**Comment by justinlittman** [comment]
Also, need to consider what will happen if stream harvester fails. Will it just get restarted by supervisor?  Will the error be reported back?

**Comment by kerchner** [comment]
Fixed by #444
